rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+map into LLVM's non-atomic volatile operations, unnecessarily differ from the
+memory access semantics of mainstream hardware in two major ways:
+
+1. Concurrent use of volatile memory operations on a given memory location is
+   considered to be Undefined Behavior, and may therefore result in unintended
+   compilation output if detected by the optimizer.
+2. Using an overly wide volatile load or store operation which cannot be carried
+   out by a single hardware load and store instruction will not result in a
+   compilation error, but in the silent emission of multiple hardware load or
+   store instructions, which might be a logic error in the users' program.
+
+By implementing support for LLVM's atomic volatile operations, and encouraging
+their use on every hardware that supports them, we eliminate these divergences
+from hardware behavior and therefore bring volatile operations closer to the
+"defer to hardware memory model" semantics that programmers expect them to have.
+This reduces the odd of mistake in programs operating outside of the regular
+Rust Abstract Machine semantics, which are notoriously hard to get right.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Rust compiler generally assumes that the program that it is building is
+living in a fully isolated memory space, where the contents of memory can only
+change if some direct action from the program (including FFI or atomic memory
+operations) allows it to change.
+
+It leverages this knowledge to transform said program's memory access patterns
+for performance optimization purposes, under the assumption that said
+transformations will not have any externally observable effect other than
+speeding up the program.
+
+Examples of such transformations include:
+
+- caching data from RAM into CPU registers,
+- eliminating unused loads and unobservable stores,
+- merging neighboring stores with each other,
+- only updating the part of an over-written struct that has actually changed.
+
+Although these optimizations are most of the time correct and useful, there are
+some situations where they are inappropriate, in areas such as:
+
+- [Memory-mapped I/O](https://en.wikipedia.org/wiki/Memory-mapped_I/O), a common
+  low-level communication protocol between CPUs and peripherals, where hardware
+  registers masquerading as memory can be used to program peripherals by
+  accessing said registers in very specific load/store patterns.
+- [Shared memory](https://en.wikipedia.org/wiki/Shared_memory), a form of
+  inter-process communication where two programs can communicate via a common
+  memory block, which means that stores are externally observable and loads are
+  not guaranteed to return consistent results.
+- Advanced uses of [virtual memory](https://en.wikipedia.org/wiki/Virtual_memory)
+  where the mere action of reading from or writing to memory may trigger
+  execution of arbitrary code by the operating system.
```

Comment:
Maybe add a remark here like "(not to be confused with shared-memory concurrency, which refers to memory being shared between multiple threads running in the same process)". This dual-use of the same term still confuses me frequently.^^

Reply:
(I assume you meant the previous bullet point, the one about "Shared memory"?)

Reply:
Ah, yes... I could have sworn I added the comment to the last line of that but it seems I slipped. Yes I meant to add this when the term "shared memory" is first used here.

Reply:
Ack, will try to find a suitable wording.

Reply:
Disambiguated.